server/quota/quota_manager.go

```diff
@@ -0,0 +1,291 @@
+package quota
+
+import (
+	"context"
+	"math"
+	"strings"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/tables"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/throttled/throttled/v2"
+	"github.com/throttled/throttled/v2/store/goredisstore.v8"
+)
+
+const (
+	// The maximum number of attempts to update rate limit data in redis.
+	maxRedisRetry = 3
+
+	// The name of the default bucket in the namespace. If you change this field,
+	// you also need to update the QuotaBucket and QuotaGroup table.
+	defaultBucketName = "default"
+
+	// THe prefix we use to all quota-related redis entries.
+	redisQuotaKeyPrefix = "quota"
+)
+
+type namespaceConfig struct {
+	bucketsByName         map[string]*tables.QuotaBucket
+	quotaKeysByBucketName map[string][]string
+}
+
+func fetchConfigFromDB(env environment.Env) (map[string]*namespaceConfig, error) {
+	if env.GetDBHandle() == nil {
+		return nil, status.FailedPreconditionError("quota manager is not configured")
+	}
+	ctx := env.GetServerContext()
+	db := env.GetDBHandle().DB(ctx)
+	bucketRows, err := db.Raw(`SELECT * FROM QuotaBuckets`).Rows()
+	if err != nil {
+		return nil, status.InternalErrorf("failed to read table QuotaBuckets: %s", err)
+	}
+	defer bucketRows.Close()
+
+	config := make(map[string]*namespaceConfig)
+	for bucketRows.Next() {
+		var tb tables.QuotaBucket
+		if err := db.ScanRows(bucketRows, &tb); err != nil {
+			return nil, status.InternalErrorf("failed to scan QuotaBuckets: %s", err)
+		}
+		if err := validateBucket(&tb); err != nil {
+			return nil, status.InternalErrorf("invalid bucket: %v", tb)
+		}
+		ns := config[tb.Namespace]
+		if ns == nil {
+			ns = &namespaceConfig{
+				bucketsByName:         make(map[string]*tables.QuotaBucket),
+				quotaKeysByBucketName: make(map[string][]string),
+			}
+			config[tb.Namespace] = ns
+		}
+		ns.bucketsByName[tb.Name] = &tb
+	}
+
+	groupRows, err := db.Raw(`SELECT * FROM QuotaGroups`).Rows()
```

Comment:
would be nice if you did both of these queries in the same transaction since you're joining the data